native/membrane_videocompositor/src/compositor/videos.rs

```diff
@@ -68,22 +81,45 @@ impl InputVideo {
 
         Self {
             yuv_textures,
-            rgba_texture,
+            frames,
             vertices,
             indices,
             position,
+            previous_frame: None,
+            single_texture_bind_group_layout,
         }
     }
 
     pub fn upload_data(
-        &self,
+        &mut self,
         device: &wgpu::Device,
         queue: &wgpu::Queue,
         converter: &YUVToRGBAConverter,
         data: &[u8],
+        pts: u64,
+        last_rendered_pts: Option<u64>,
     ) {
         self.yuv_textures.upload_data(queue, data);
-        converter.convert(device, queue, &self.yuv_textures, &self.rgba_texture);
+        let frame = RGBATexture::new(
+            device,
+            self.position.width,
+            self.position.height,
+            &self.single_texture_bind_group_layout,
+        );
+        converter.convert(device, queue, &self.yuv_textures, &frame);
+
+        // if we haven't rendered a frame yet, or pts of our frame is ahead of last rendered frame
+        if last_rendered_pts.is_none() || pts > last_rendered_pts.unwrap() {
```

Comment:
I'd really like to see it in the separate private method with a meaningfull name.

Reply:
as discussed I'll leave this as is